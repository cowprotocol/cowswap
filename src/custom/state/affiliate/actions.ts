import { createAction } from '@reduxjs/toolkit'

export const updateReferralAddress = createAction<{
  value: string
  isValid: boolean
} | null>('affiliate/updateReferralAddress')

export const setReferralAddressActive = createAction<boolean>('affiliate/setReferralAddressActive')

export const updateAddress = createAction<string>('affiliate/updateAddress')

export const dismissNotification = createAction<string>('affiliate/dismissNotification')
