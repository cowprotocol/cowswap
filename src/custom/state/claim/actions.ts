import { createAction } from '@reduxjs/toolkit'
import { SupportedChainId } from 'constants/chains'
import { ClaimInfo } from 'state/claim/reducer'

export enum ClaimStatus {
  DEFAULT = 'DEFAULT',
  ATTEMPTING = 'ATTEMPTING',
  SUBMITTED = 'SUBMITTED',
  CONFIRMED = 'CONFIRMED',
  FAILED = 'FAILED',
}

export type ClaimActions = {
  // account
  setInputAddress: (payload: string) => void
  setActiveClaimAccount: (payload: string) => void
  setActiveClaimAccountENS: (payload: string) => void

  // search
  setIsSearchUsed: (payload: boolean) => void

  // claiming
  setClaimStatus: (payload: ClaimStatus) => void
  setClaimedAmount: (payload: string) => void
  setEstimatedGas: (payload: string) => void

  // investing
  setIsInvestFlowActive: (payload: boolean) => void
  setInvestFlowStep: (payload: number) => void
  initInvestFlowData: () => void
  updateInvestAmount: (payload: { index: number; amount: string }) => void
  updateInvestError: (payload: { index: number; error: string | undefined }) => void
  setIsTouched: (payload: { index: number; isTouched: boolean }) => void

  // claim row selection
  setSelected: (payload: number[]) => void
  setSelectedAll: (payload: boolean) => void
}

// accounts
export const setInputAddress = createAction<string>('claim/setInputAddress')
export const setActiveClaimAccount = createAction<string>('claim/setActiveClaimAccount')
export const setActiveClaimAccountENS = createAction<string>('claim/setActiveClaimAccountENS')

// search
export const setIsSearchUsed = createAction<boolean>('claim/setIsSearchUsed')

// claiming
export const setClaimedAmount = createAction<string>('claim/setClaimedAmount')
export const setClaimStatus = createAction<ClaimStatus>('claim/setClaimStatus')
export const setEstimatedGas = createAction<string>('claim/setEstimatedGas')

// investing
export const setIsInvestFlowActive = createAction<boolean>('claim/setIsInvestFlowActive')
export const setInvestFlowStep = createAction<number>('claim/setInvestFlowStep')
export const initInvestFlowData = createAction('claim/initInvestFlowData')
export const updateInvestAmount = createAction<{
  index: number
  amount: string
}>('claim/updateInvestAmount')
export const updateInvestError = createAction<{
  index: number
  error: string | undefined
}>('claim/updateInvestError')
export const setIsTouched = createAction<{
  index: number
  isTouched: boolean
}>('claim/setIsTouched')
// claim row selection
export const setSelected = createAction<number[]>('claim/setSelected')
export const setSelectedAll = createAction<boolean>('claim/setSelectedAll')
// Claim UI reset sugar
export const resetClaimUi = createAction('claims/resetClaimUi')
// Claims on other chains
export const setClaimsCount = createAction<{
  chain: SupportedChainId
  claimInfo: Partial<ClaimInfo>
  account: string
}>('claims/setClaimsCount')
