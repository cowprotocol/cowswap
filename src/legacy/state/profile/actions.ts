import { createAction } from '@reduxjs/toolkit'

export const closeAnnouncement = createAction<{ contentHash: string }>('profile/closeAnnouncement')
