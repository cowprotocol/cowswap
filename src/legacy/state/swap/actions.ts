import { createAction } from '@reduxjs/toolkit'

export enum Field {
  INPUT = 'INPUT',
  OUTPUT = 'OUTPUT',
}

export interface ReplaceOnlyTradeRawStatePayload {
  readonly chainId: number | null
  readonly inputCurrencyId?: string
  readonly outputCurrencyId?: string
  readonly recipient: string | null
}

export interface ReplaceSwapStatePayload extends ReplaceOnlyTradeRawStatePayload {
  readonly typedValue: string
  readonly independentField: Field
}

export const selectCurrency = createAction<{ field: Field; currencyId: string }>('swap/selectCurrency')
export const switchCurrencies = createAction<void>('swap/switchCurrencies')
export const typeInput = createAction<{ field: Field; typedValue: string }>('swap/typeInput')

/**
 * Replaces only a subset of the state corresponding to TradeRawState.
 */
export const replaceOnlyTradeRawState = createAction<ReplaceOnlyTradeRawStatePayload>('swap/replaceOnlyTradeRawState')
export const replaceSwapState = createAction<ReplaceSwapStatePayload>('swap/replaceSwapState')
export const setRecipient = createAction<{ recipient: string | null }>('swap/setRecipient')
