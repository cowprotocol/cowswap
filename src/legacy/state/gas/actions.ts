import { createAction } from '@reduxjs/toolkit'
import { WithChainId } from 'legacy/state/lists/actions'
import { GasFeeEndpointResponse } from 'api/gasPrices'

export type UpdateGasPrices = GasFeeEndpointResponse & WithChainId

export const updateGasPrices = createAction<UpdateGasPrices>('gas/updateGasPrices')
