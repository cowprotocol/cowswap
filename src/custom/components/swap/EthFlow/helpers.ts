import { useState } from 'react'
import { parseUnits } from '@ethersproject/units'
import { CurrencyAmount, Currency } from '@uniswap/sdk-core'
import { BigNumber } from '@ethersproject/bignumber'
// eslint-disable-next-line no-restricted-imports
import { t } from '@lingui/macro'
import { useGasPrices } from 'state/gas/hooks'
import { PendingHashMap, Props } from 'components/swap/EthFlow'
import { ActivityDerivedState } from 'components/AccountDetails/Transaction'
import { useSingleActivityState } from 'hooks/useActivityDerivedState'
import { ApprovalState } from 'hooks/useApproveCallback'

export const MINIMUM_TXS = '10'
export const AVG_APPROVE_COST_GWEI = '50000'
export const DEFAULT_GAS_FEE = parseUnits('50', 'gwei')

export const _setNativeLowBalanceError = (nativeSymbol: string) =>
  new Error(
    t`This ${nativeSymbol} wrapping operation may leave insufficient funds to cover any future on-chain transaction costs.`
  )

export function _isLowBalanceCheck({
  threshold,
  txCost,
  nativeInput,
  balance,
}: {
  threshold?: CurrencyAmount<Currency>
  txCost?: CurrencyAmount<Currency>
  nativeInput?: CurrencyAmount<Currency>
  balance?: CurrencyAmount<Currency>
}) {
  if (!threshold || !txCost) return false
  if (!nativeInput || !balance || nativeInput.add(txCost).greaterThan(balance)) return true
  // OK if: users_balance - (amt_input + 1_tx_cost) > low_balance_threshold
  return balance.subtract(nativeInput.add(txCost)).lessThan(threshold)
}

export const _getAvailableTransactions = ({
  nativeBalance,
  nativeInput,
  singleTxCost,
}: {
  nativeBalance?: CurrencyAmount<Currency>
  nativeInput?: CurrencyAmount<Currency>
  singleTxCost?: CurrencyAmount<Currency>
}) => {
  if (!nativeBalance || !nativeInput || !singleTxCost || nativeBalance.lessThan(nativeInput.add(singleTxCost))) {
    return null
  }

  // USER_BALANCE - (USER_WRAP_AMT + 1_TX_CST) / 1_TX_COST = AVAILABLE_TXS
  const txsAvailable = nativeBalance.subtract(nativeInput.add(singleTxCost)).divide(singleTxCost)
  return txsAvailable.lessThan('1') ? null : txsAvailable.quotient.toString()
}

export function _estimateTxCost(gasPrice: ReturnType<typeof useGasPrices>, native: Currency | undefined) {
  if (!native) {
    return {}
  }
  // TODO: should use DEFAULT_GAS_FEE from backup source
  // when/if implemented
  const gas = gasPrice?.standard || DEFAULT_GAS_FEE

  const amount = BigNumber.from(gas).mul(MINIMUM_TXS).mul(AVG_APPROVE_COST_GWEI)

  return {
    multiTxCost: CurrencyAmount.fromRawAmount(native, amount.toString()),
    singleTxCost: CurrencyAmount.fromFractionalAmount(native, amount.toString(), MINIMUM_TXS),
  }
}

export enum EthFlowState {
  SwapReady, // 0
  WrapAndApproveNeeded, // 1
  WrapAndApprovePending, // 2
  ApproveNeeded, // 3
  WrapNeeded, // 4
  ApprovePending, // 5
  ApproveInsufficient, // 6
  ApproveFailed, // 7
  WrapFailed, // 8
  WrapPending, // 9
  WrapAndApproveFailed, // 10
  Loading, // 11
}

export type DerivedEthFlowStateProps = Pick<ModalTextContentProps, 'isExpertMode'> & {
  approveError: Error | null
  wrapError: Error | null
  approveState: ActivityDerivedState | null
  wrapState: ActivityDerivedState | null
  needsApproval: boolean | undefined
  needsWrap: boolean | undefined
}

// returns derived ethflow state from current props
export function _getDerivedEthFlowState(params: DerivedEthFlowStateProps) {
  const { approveError, wrapError, approveState, wrapState, needsApproval, needsWrap, isExpertMode } = params
  // approve state
  const approveExpired = approveState?.isExpired
  const approvePending = approveState?.isPending
  const approveSentAndSuccessful = Boolean(!approveError && !approvePending && approveState?.isConfirmed)
  const approveInsufficient = approveSentAndSuccessful && needsApproval
  const approveFinished = !needsApproval || approveSentAndSuccessful
  // wrap state
  const wrapExpired = wrapState?.isExpired
  const wrapPending = wrapState?.isPending
  const wrapSentAndSuccessful = Boolean(!wrapError && !wrapPending && wrapState?.isConfirmed)
  const wrapNeeded = needsWrap && !wrapSentAndSuccessful
  const wrapFinished = !needsWrap || wrapSentAndSuccessful

  // PENDING states
  if (wrapPending || approvePending) {
    // expertMode only - both operations pending
    if (isExpertMode && wrapPending && approvePending) {
      return EthFlowState.WrapAndApprovePending
    }
    // Only wrap is pending
    else if (wrapPending) {
      return EthFlowState.WrapPending
    }
    // Only approve is pending
    else return EthFlowState.ApprovePending
  }
  // FAILED states
  else if (approveExpired || wrapExpired) {
    // expertMode only - BOTH operations failed
    if (isExpertMode && approveExpired && wrapExpired) {
      return EthFlowState.WrapAndApproveFailed
    }
    // Only wrap failed
    else if (wrapExpired) {
      return EthFlowState.WrapFailed
    }
    // Only approve failed
    else return EthFlowState.ApproveFailed
  }
  // NEEDS wrap/approve state
  else if (needsApproval || wrapNeeded) {
    // INSUFFICIENT approve state
    if (approveInsufficient) {
      return EthFlowState.ApproveInsufficient
    }
    // in expertMode and we need to wrap and swap
    else if (isExpertMode && needsApproval && wrapNeeded) {
      return EthFlowState.WrapAndApproveNeeded
    }
    // Only wrap needed
    else if (wrapNeeded) {
      return EthFlowState.WrapNeeded
    }
    // Only approve needed
    else {
      return EthFlowState.ApproveNeeded
    }
  }
  // BOTH successful, ready to swap
  else if (approveFinished && wrapFinished) {
    return EthFlowState.SwapReady
  }
  // LOADING
  else {
    return EthFlowState.Loading
  }
}

export type ModalTextContentProps = {
  wrappedSymbol: string
  nativeSymbol: string
  state: EthFlowState
  isExpertMode: boolean
  isNative: boolean
  wrapSubmitted: boolean
  approveSubmitted: boolean
}

// returns modal content: header and descriptions based on state
export function _getModalTextContent(params: ModalTextContentProps) {
  const { wrappedSymbol, nativeSymbol, state, isExpertMode, isNative /*, wrapSubmitted, approveSubmitted */ } = params
  // common text
  const swapHeader = `Swap ${wrappedSymbol}`
  const swapDescription = `Click "Swap" to submit an off-chain transaction and swap your ${wrappedSymbol}`
  const signatureRequiredDescription = 'Transaction signature required, please check your connected wallet'
  const transactionInProgress = 'Transaction in progress. See below for live status updates'
  // wrap
  const wrapHeader = isNative ? `Wrap your ${nativeSymbol}` : `Unwrap your ${wrappedSymbol}`
  const wrapInstructions = `Submit an on-chain ${isNative ? 'wrap' : 'unwrap'} transaction to convert your ${
    isNative ? nativeSymbol : wrappedSymbol
  } into ${isNative ? wrappedSymbol : nativeSymbol}`
  // approve
  const approveHeader = `Approve ${wrappedSymbol}`
  const approveInstructions = `Give CoW Protocol permission to swap your ${wrappedSymbol} via an on-chain ERC20 Approve transaction`
  // both
  const bothHeader = `Wrap and approve`
  const bothInstructions = `2 pending on-chain transactions: ${
    isNative ? `${nativeSymbol} wrap` : `${wrappedSymbol} unwrap`
  } and approve. Please check your connected wallet for both signature requests`

  let header = ''
  let description: string | null = null
  switch (state) {
    /**
     * FAILED operations
     * wrap/approve/both in expertMode failed
     */
    case EthFlowState.WrapAndApproveFailed: {
      header = 'Wrap and Approve failed!'
      description =
        'Both wrap and approve operations failed. Check that you are providing a sufficient gas limit for both transactions in your wallet. Click "Wrap and approve" to try again'
      break
    }
    case EthFlowState.WrapFailed: {
      header = `Wrap ${nativeSymbol} failed!`
      description = `Wrap operation failed. Check that you are providing a sufficient gas limit for the transaction in your wallet. Click "Wrap ${nativeSymbol}" to try again`
      break
    }
    case EthFlowState.ApproveFailed: {
      header = `Approve ${wrappedSymbol} failed!`
      description = `Approve operation failed. Check that you are providing a sufficient gas limit for the transaction in your wallet. Click "Approve ${wrappedSymbol}" to try again`
      break
    }

    /**
     * PENDING operations
     * wrap/approve/both in expertMode
     */
    case EthFlowState.WrapAndApprovePending: {
      header = bothHeader
      description = 'Transactions in progress. See below for live status updates of each operation'
      break
    }
    case EthFlowState.WrapPending:
    case EthFlowState.ApprovePending: {
      description = transactionInProgress
      // wrap only
      if (state === EthFlowState.WrapPending) {
        header = wrapHeader
      }
      // approve only
      else {
        header = approveHeader
      }
      break
    }

    case EthFlowState.ApproveInsufficient: {
      header = 'Approval amount insufficient!'
      description =
        'Approval amount insufficient for input amount. Check that you are approving an amount equal to or greater than the input amount'
      break
    }

    /**
     * NEEDS operations
     * need to wrap/approve/both in expertMode
     */
    case EthFlowState.WrapAndApproveNeeded: {
      header = bothHeader
      description = bothInstructions
      break
    }
    case EthFlowState.WrapNeeded:
    case EthFlowState.ApproveNeeded: {
      // wrap only
      if (state === EthFlowState.WrapNeeded) {
        header = wrapHeader
        description = wrapInstructions
      }
      // approve only
      else {
        header = approveHeader
        description = approveInstructions
      }

      // in expert mode tx popups are automatic
      // so we show user message to check wallet popup
      if (isExpertMode) {
        description = signatureRequiredDescription
      }
      break
    }

    /**
     * SWAP operation ready
     */
    case EthFlowState.SwapReady: {
      header = swapHeader
      description = isExpertMode ? null : swapDescription
      break
    }

    // show generic operation loading as default
    // to shut TS up
    default: {
      header = 'Loading operation'
      description = 'Operation in progress!'
      break
    }
  }

  return { header, description }
}

// returns proper prop for visualiser: which currency is shown on left vs right (wrapped vs unwrapped)
export function _getCurrencyForVisualiser<T>(native: T, wrapped: T, isWrap: boolean, isUnwrap: boolean) {
  if (isWrap || isUnwrap) {
    return isWrap ? native : wrapped
  } else {
    return native
  }
}

export type ActionButtonParams = Pick<Props, 'isNativeIn'> &
  Pick<DerivedEthFlowStateProps, 'approveError' | 'wrapError' | 'approveState' | 'wrapState' | 'isExpertMode'> &
  Pick<ModalTextContentProps, 'nativeSymbol' | 'wrappedSymbol' | 'state'> & {
    isWrap: boolean
    loading: boolean
    handleSwap: (showSwapModal?: boolean) => Promise<void>
    handleApprove: () => Promise<void>
    handleWrap: () => Promise<void>
    handleMountInExpertMode: () => Promise<void>
  }
// conditionally renders the correct action button depending on the proposed action and current eth-flow state
export function _getActionButtonProps({
  approveError,
  wrapError,
  approveState,
  wrapState,
  isNativeIn,
  nativeSymbol,
  wrappedSymbol,
  isExpertMode,
  state,
  isWrap,
  loading,
  handleSwap,
  handleWrap,
  handleApprove,
  handleMountInExpertMode,
}: ActionButtonParams) {
  // async, pre-receipt errors (e.g user rejected TX)
  const hasErrored = !!(approveError || wrapError)

  let label = ''
  let showButton = !isExpertMode
  let showLoader = loading
  // dynamic props for cta button
  const buttonProps: {
    disabled: boolean
    onClick: (() => Promise<void>) | undefined
  } = {
    disabled: false,
    onClick: undefined,
  }

  switch (state) {
    // an operation has failed after submitting
    case EthFlowState.WrapAndApproveFailed:
      label = 'Wrap and approve'
      showButton = true
      buttonProps.onClick = handleMountInExpertMode
      // disable button on load (after clicking)
      buttonProps.disabled = showLoader
      break
    case EthFlowState.WrapFailed:
      label = `Wrap ${nativeSymbol}`
      showButton = true
      buttonProps.onClick = handleWrap
      break
    case EthFlowState.ApproveFailed:
      label = `Approve ${wrappedSymbol}`
      showButton = true
      buttonProps.onClick = handleApprove
      break
    // non failures
    case EthFlowState.WrapNeeded:
      label = isNativeIn || isWrap ? 'Wrap ' + nativeSymbol : 'Unwrap ' + wrappedSymbol
      buttonProps.onClick = handleWrap
      buttonProps.disabled = Boolean(wrapState?.isPending)
      break
    case EthFlowState.ApproveNeeded:
    case EthFlowState.ApproveInsufficient:
      label = 'Approve ' + wrappedSymbol
      // Show button if approve insufficient (applies to expertMode)
      if (state === EthFlowState.ApproveInsufficient) {
        showButton = true
      }
      buttonProps.onClick = handleApprove
      buttonProps.disabled = Boolean(approveState?.isPending)
      break
    case EthFlowState.SwapReady:
      label = 'Swap'
      buttonProps.onClick = () => handleSwap(true)
      buttonProps.disabled = loading || hasErrored
      break
    // loading = default
    default:
      buttonProps.disabled = true
      showLoader = true
      break
  }

  return {
    label,
    showButton,
    showLoader,
    buttonProps,
  }
}

/**
 * useEthFlowStatesAndSetters
 *
 * @returns all ETH-FLOW states and setters related to wrap/approve
 */
export function useEthFlowStatesAndSetters({
  chainId,
  approvalState,
  wrapState,
}: Pick<Props, 'approvalState' | 'wrapState'> & {
  chainId?: number
}) {
  const [pendingHashMap, setPendingHashMap] = useState<PendingHashMap>({
    approveHash: undefined,
    wrapHash: undefined,
  })
  // maintain own local state of approve/wrap states
  const [loading, setLoading] = useState(false)
  // APPROVE STATE - use activity state and derive isPending based on both the hook and activity state
  const approvalActivityState = useSingleActivityState({ chainId, id: pendingHashMap.approveHash || '' })
  const approvalDerivedState = !!approvalActivityState
    ? {
        ...approvalActivityState,
        isPending: approvalActivityState?.isPending || approvalState === ApprovalState.PENDING,
      }
    : null
  const [approveSubmitted, setApproveSubmitted] = useState(false)
  const [approveError, setApproveError] = useState<Error | null>(null)
  // WRAP STATE
  const wrapActivityState = useSingleActivityState({ chainId, id: pendingHashMap.wrapHash || '' })
  const [wrapSubmitted, setWrapSubmitted] = useState(false)
  const [wrapError, setWrapError] = useState<Error | null>(null)

  return {
    pendingHashMap,
    setPendingHashMap,
    loading,
    setLoading,
    // APPROVE
    approvalDerivedState,
    approveSubmitted,
    setApproveSubmitted,
    approveError,
    setApproveError,
    // WRAPPING
    wrapDerivedState: wrapActivityState,
    wrapSubmitted,
    setWrapSubmitted,
    wrapError,
    setWrapError,
  }
}
