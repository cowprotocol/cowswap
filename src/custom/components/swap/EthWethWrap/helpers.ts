import { parseUnits } from '@ethersproject/units'
import { CurrencyAmount, Currency } from '@uniswap/sdk-core'
import { BigNumber } from '@ethersproject/bignumber'
// eslint-disable-next-line no-restricted-imports
import { t } from '@lingui/macro'
import { useGasPrices } from 'state/gas/hooks'

export const MINIMUM_TXS = '10'
export const AVG_APPROVE_COST_GWEI = '50000'
export const DEFAULT_GAS_FEE = parseUnits('50', 'gwei')

export const _setNativeLowBalanceError = (nativeSymbol: string) =>
  new Error(
    t`This ${nativeSymbol} wrapping operation may leave insufficient funds to cover any future on-chain transaction costs.`
  )

export function _isLowBalanceCheck({
  threshold,
  txCost,
  nativeInput,
  balance,
}: {
  threshold?: CurrencyAmount<Currency>
  txCost?: CurrencyAmount<Currency>
  nativeInput?: CurrencyAmount<Currency>
  balance?: CurrencyAmount<Currency>
}) {
  if (!threshold || !txCost) return false
  if (!nativeInput || !balance || nativeInput.add(txCost).greaterThan(balance)) return true
  // OK if: users_balance - (amt_input + 1_tx_cost) > low_balance_threshold
  return balance.subtract(nativeInput.add(txCost)).lessThan(threshold)
}

export const _getAvailableTransactions = ({
  nativeBalance,
  nativeInput,
  singleTxCost,
}: {
  nativeBalance?: CurrencyAmount<Currency>
  nativeInput?: CurrencyAmount<Currency>
  singleTxCost?: CurrencyAmount<Currency>
}) => {
  if (!nativeBalance || !nativeInput || !singleTxCost || nativeBalance.lessThan(nativeInput.add(singleTxCost))) {
    return null
  }

  // USER_BALANCE - (USER_WRAP_AMT + 1_TX_CST) / 1_TX_COST = AVAILABLE_TXS
  const txsAvailable = nativeBalance.subtract(nativeInput.add(singleTxCost)).divide(singleTxCost)
  return txsAvailable.lessThan('1') ? null : txsAvailable.quotient.toString()
}

export function _estimateTxCost(gasPrice: ReturnType<typeof useGasPrices>, native: Currency | undefined) {
  if (!native) {
    return {}
  }
  // TODO: should use DEFAULT_GAS_FEE from backup source
  // when/if implemented
  const gas = gasPrice?.standard || DEFAULT_GAS_FEE

  const amount = BigNumber.from(gas).mul(MINIMUM_TXS).mul(AVG_APPROVE_COST_GWEI)

  return {
    multiTxCost: CurrencyAmount.fromRawAmount(native, amount.toString()),
    singleTxCost: CurrencyAmount.fromFractionalAmount(native, amount.toString(), MINIMUM_TXS),
  }
}
