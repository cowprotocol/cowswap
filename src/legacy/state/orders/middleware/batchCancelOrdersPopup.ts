import { Dispatch, MiddlewareAPI } from 'redux'

import { orderAnalytics } from '../../../components/analytics'
import { addPopup, AddPopupPayload } from '../../application/reducer'
import { AppState } from '../../index'
import { BaseOrder } from '../actions'
import { buildCancellationPopupSummary } from '../buildCancellationPopupSummary'
import { OrderTxTypes, setPopupData } from '../helpers'

export function batchCancelOrdersPopup(store: MiddlewareAPI<Dispatch, AppState>, orders: BaseOrder[]) {
  // construct Cancelled Order Popups for each Order
  orders.forEach((order) => {
    const popup = _buildCancellationPopup(order)
    orderAnalytics('Canceled', order.class)

    store.dispatch(addPopup(popup))
  })
}

function _buildCancellationPopup(order: BaseOrder) {
  const { cancellationHash, apiAdditionalInfo, id, summary } = order

  if (cancellationHash && !apiAdditionalInfo) {
    // EthFlow order which has been cancelled and does not exist on the backend
    // Use the `tx` popup
    return setPopupData(OrderTxTypes.TXN, {
      success: true,
      summary: buildCancellationPopupSummary(id, summary),
      hash: cancellationHash,
      id,
    }) as unknown as AddPopupPayload
  } else {
    // Regular order being cancelled
    // Use `metatx` popup
    return setPopupData(OrderTxTypes.METATXN, {
      success: true,
      summary: buildCancellationPopupSummary(id, summary),
      id,
    }) as unknown as AddPopupPayload
  }
}
