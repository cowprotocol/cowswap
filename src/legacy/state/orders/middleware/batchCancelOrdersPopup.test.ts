import { OrderClass } from '@cowprotocol/cow-sdk'

import { AnyAction, Dispatch, MiddlewareAPI } from 'redux'
import { anything, capture, instance, mock, resetCalls, verify, when } from 'ts-mockito'

import { batchCancelOrdersPopup } from './batchCancelOrdersPopup'

import { AppState } from '../../index'
import { CancelOrdersBatchParams } from '../actions'
import { setPopupData } from '../helpers'

const MOCK_ETHFLOW_ORDER = {
  '0x001': {
    id: '0x001',
    order: {
      id: '0x001',
      summary: 'summary',
      orderClass: OrderClass.MARKET,
      orderCreationHash: '0xhash',
    },
  },
}

const MOCK_REGULAR_ORDER = {
  '0x002': {
    id: '0x002',
    order: {
      id: '0x002',
      summary: 'summary',
      orderClass: OrderClass.LIMIT,
      cancellationHash: '0xhash',
    },
  },
}

const MOCK_ORDERS_STORE = {
  pending: { ...MOCK_ETHFLOW_ORDER, ...MOCK_REGULAR_ORDER },
}

const MOCK_POPUP_DATA = 'mock popup data'

jest.mock('../helpers', () => ({
  ...jest.requireActual('../helpers'),
  setPopupData: jest.fn(),
}))

const setPopupDataMock = setPopupData as jest.MockedFunction<typeof setPopupData>
const storeMock = mock<MiddlewareAPI<Dispatch, AppState>>()
const payloadMock = mock<CancelOrdersBatchParams>()

describe('batchCancelOrdersPopup', () => {
  beforeEach(() => {
    resetCalls(storeMock)
    resetCalls(payloadMock)
    setPopupDataMock.mockReturnValue(MOCK_POPUP_DATA as any)
  })

  it('should not trigger pop up if there are no pending orders', () => {
    when(payloadMock.ids).thenReturn(['0x000'])

    // @ts-ignore
    batchCancelOrdersPopup(instance(storeMock), instance(payloadMock), MOCK_ORDERS_STORE)

    verify(storeMock.dispatch(anything())).never()
  })

  it('should trigger pop ups if there are pending orders', () => {
    when(payloadMock.ids).thenReturn(['0x001', '0x002'])

    // @ts-ignore
    batchCancelOrdersPopup(instance(storeMock), instance(payloadMock), MOCK_ORDERS_STORE)

    const [addPopupAction] = capture(storeMock.dispatch<AnyAction>).first()

    expect(addPopupAction.payload).toEqual(MOCK_POPUP_DATA)

    verify(storeMock.dispatch(anything())).twice()
  })
})
