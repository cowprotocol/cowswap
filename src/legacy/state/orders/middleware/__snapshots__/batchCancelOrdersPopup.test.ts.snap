// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batchCancelOrdersPopup should trigger pop ups if there are pending orders 1`] = `
Object {
  "payload": Object {
    "id": "0x001",
    "popup": Object {
      "content": Object {
        "metatxn": Object {
          "id": "0x001",
          "success": true,
          "summary": <buildCancellationPopupSummary__Wrapper>
            <p>
              Order successfully cancelled
            </p>
            <p>
              Order 
              <strong>
                0x001
              </strong>
              :
            </p>
            <styled__CancellationSummary
              as="p"
            >
              summary
            </styled__CancellationSummary>
          </buildCancellationPopupSummary__Wrapper>,
        },
      },
      "key": "0x001_undefined",
    },
  },
  "type": "application/addPopup",
}
`;
